haskell/cabal.bzl

```diff
@@ -892,23 +892,71 @@ def _get_components(components, package):
     """
     return components.get(package, _default_components.get(package, struct(lib = True, exe = [])))
 
-def _validate_package_specs(package_specs):
-    found_ty = type(package_specs)
-    if found_ty != "list":
-        fail("Unexpected output format for `stack ls dependencies json`. Expected 'list', but got '%s'." % found_ty)
-
-def _validate_package_spec(package_spec):
-    fields = [
-        ("name", "string"),
-        ("version", "string"),
-        ("dependencies", "list"),
-    ]
-    for (field, ty) in fields:
-        if not field in package_spec:
-            fail("Unexpected output format for `stack ls dependencies json`. Missing field '%s'." % field)
-        found_ty = type(package_spec[field])
-        if found_ty != ty:
-            fail("Unexpected output format for `stack ls dependencies json`. Expected field '%s' of type '%s', but got type '%s'." % (field, ty, found_ty))
+def _parse_json_field(json, field, ty, fmt):
+    """Read and type-check a field from a JSON object.
+
+    Invokes `fail` on error.
+
+    Attrs:
+      json: dict, The parsed JSON object.
+      field: string, The name of the field.
+      ty: string, The expected type of the field.
+      fmt: string, Error message format string. E.g. `Error: %s`.
+
+    Returns:
+      The value of the field.
+    """
+    if not field in json:
+        fail(fmt % "Missing field '{field}'.".format(field = field))
+    actual_ty = type(json[field])
+    if actual_ty != ty:
+        fail(fmt % "Expected field '{field}' of type '{expected}', but got '{got}'.".format(
+            field = field,
+            expected = ty,
+            got = actual_ty,
+        ))
+    return json[field]
+
+def _parse_package_spec(package_spec):
+    errmsg = "Unexpected output format for `stack ls dependencies json` in {context}: %s"
+
+    # Parse simple fields.
+    parsed = {
+        field: _parse_json_field(package_spec, field, ty, errmsg.format(context = "package description"))
+        for (field, ty) in [("name", "string"), ("version", "string"), ("dependencies", "list")]
+    }
+
+    # Parse location field.
+    location = {}
+    if parsed["name"] in _CORE_PACKAGES or not "location" in package_spec:
+        location["type"] = "core"
+    else:
+        location_type = _parse_json_field(package_spec["location"], "type", "string", errmsg.format(context = "location description"))
+        if location_type == "project package":
+            location["type"] = "vendored"
+        elif location_type == "hackage":
+            location["type"] = location_type
+            url_prefix = _parse_json_field(package_spec["location"], "url", "string", errmsg.format(context = location_type + " location description"))
+            location["url"] = url_prefix + "/{name}-{version}.tar.gz".format(**parsed)
+            # stack does not expose sha-256, see https://github.com/commercialhaskell/stack/issues/5274
```

Comment:
why is that important?

Reply:
Once we download Hackage and archive dependencies using Bazel rather than `stack unpack` we'll want the SHA-256 to make sure the download is reproducible and caches well in Bazel's repository cache. However, since `stack ls dependencies json` doesn't expose the SHA-256 we have to determine it in other ways. https://github.com/tweag/rules_haskell/pull/1376 implements that.

If `stack` implements https://github.com/commercialhaskell/stack/issues/5274 then we can directly extract the SHA-256 here instead of determining it ourselves later on.